GMCM support for CP packs (not CP itself though)

spacechase0 opened the pull request:
I didn't entirely understand the changes I needed to make to `ManualValueProvider`, so let me know if those are off.

Tested with Lumisteria Flowery Grass.

spacechase0 committed:
GMCM support for CP packs (not CP itself though)

Pathoschild commented on ContentPatcher/ModEntry.cs:
```diff
@@ -356,15 +357,43 @@ private void LoadContentPacks(IEnumerable<RawContentPack> contentPacks, Invarian
+                        {
                             this.Monitor.VerboseLog($"   found config.json with {config.Count} fields...");
 
+                            configMenuIntegration = new GenericModConfigMenuIntegration<InvariantDictionary<ConfigField>>(
```
I try to keep integration logic out of the main mod code. Can you move the new code out of `ModEntry` into a separate class similar to the existing `GenericModConfigMenuIntegrationForFastAnimations`/`…Tractor` classes?

Pathoschild replied:
Done.

Pathoschild commented on ContentPatcher/ModEntry.cs:
```diff
@@ -356,15 +357,43 @@ private void LoadContentPacks(IEnumerable<RawContentPack> contentPacks, Invarian
                         {
                             ConfigField field = pair.Value;
-                            IValueProvider valueProvider = new ImmutableValueProvider(pair.Key, field.Value, allowedValues: field.AllowValues, canHaveMultipleValues: field.AllowMultiple);
+                            ManualValueProvider valueProvider = new ManualValueProvider(pair.Key);
```
This will break config tokens used in immutable fields like `Enabled`. We should probably keep these as `ImmutableValueProvider`, and remove/add the token when we change the config.

Pathoschild replied:
Done.

Pathoschild commented on ContentPatcher/Framework/ModTokenContext.cs:
```diff
@@ -146,7 +146,7 @@ public void UpdateContext(InvariantHashSet globalChangedTokens)
             foreach (ManagedManualToken managed in this.DynamicTokens.Values)
             {
-                managed.ValueProvider.SetValue(null);
+                managed.ValueProvider.SetValue((ITokenString)null);
```
This conflicts with some CP 1.17 changes; can you rebase your changes onto the latest `develop`?

Pathoschild replied:
Done.

Pathoschild commented on ContentPatcher/ModEntry.cs:
```diff
@@ -463,5 +492,71 @@ select value.Trim()
+                        Action<InvariantDictionary<ConfigField>, bool> set = (config, val) =>
+                        {
+                            InvariantHashSet values = new InvariantHashSet(configField.Value);
+                            if (val && !values.Contains(allowedValue))
```
* I suggest renaming `allowedValue` to `value`, and `val` to `selected` for clarity.
* Since we'll immediately set the `Value` field to the new set, we can just change it directly.
* We can simplify a bit, since `HashSet` will ignore duplicates and doesn't mind if you try removing a value that's not there. 

```cs
// old code
InvariantHashSet values = new InvariantHashSet(configField.Value);
if (val && !values.Contains(allowedValue))
    values.Add(allowedValue);
if (!val && values.Contains(allowedValue)) 
    values.Remove(allowedValue);
...
configField.Value = values;
```
```cs
// suggested code
if (selected)
   configField.Value.Add(value);
else
   configField.Value.Remove(value);
```

Pathoschild replied:
Done.

Pathoschild commented on ContentPatcher/ModEntry.cs:
```diff
@@ -463,5 +492,71 @@ select value.Trim()
+                                values.Remove(allowedValue);
+
+                            if (values.Count == 0 && !configField.AllowBlank)
+                                values = configField.DefaultValues;
```
We should copy `DefaultValues` here to avoid passing a reference to the original field:
```cs
values = new InvariantHashSet(configField.DefaultValues);
```

Pathoschild replied:
Done.

Pathoschild commented on ContentPatcher/ModEntry.cs:
```diff
@@ -463,5 +492,71 @@ select value.Trim()
+                    if (configField.AllowBlank)
+                        choices.Insert(0, "");
+
+                    Func<InvariantDictionary<ConfigField>, string> get = (config) => configField.Value.Count > 0 ? configField.Value.First() : "";
```
This can be simplified:
```cs
// current code
configField.Value.Count > 0 ? configField.Value.First() : ""

// suggested
configField.Value.FirstOrDefault() ?? ""
```

Pathoschild replied:
Done.

Pathoschild commented on ContentPatcher/ModEntry.cs:
```diff
@@ -463,5 +492,71 @@ select value.Trim()
+                Func<InvariantDictionary<ConfigField>, string> get = (config) => string.Join( ", ", configField.Value.ToArray());
+                Action<InvariantDictionary<ConfigField>, string> set = (config, val) =>
+                {
+                    InvariantHashSet values = configField.AllowMultiple ? this.ParseCommaDelimitedField(val) : new InvariantHashSet(val);
```
Allowing commas in a value may cause odd effects, since Content Patcher often assumes comma = separate value. Is it possible to pass a filter to GMCM so prohibited characters can't be typed instead?

spacechase0 replied:
Not at the moment. I could add a RegisterSimpleOptionFilter(IManifest manifest, string optionName, Func<string, string> filter). Would that work?

Pathoschild replied:
I changed the logic to only save the first value instead if multiple values aren't allowed, which is good enough for now and should be reasonably intuitive.

Pathoschild reviewed (changes requested):
Thanks! I added some initial code review comments inline.

Pathoschild committed:
switch configs back to immutable tokens, reregister tokens when they change instead

Pathoschild committed:
simplify code

Pathoschild committed:
fix config field's value & default values set to same reference

Pathoschild committed:
fix config UI letting you save value containing commas

This would have odd effects, since commas are used as value delimiters in Content Patcher.

Pathoschild committed:
shorten variable names

Pathoschild committed:
refactor GMCM logic into separate integration class

Pathoschild committed:
represent boolean config fields as checkboxes

Pathoschild commented:
Thanks! Refactored a bit and merged into `develop` for the upcoming Content Patcher 1.18.
